### PR TITLE
Improve inlining of `return_call*`

### PR DESCRIPTION
**Pull request by vouillon**
Use the previous implementation when no `return_call` is in a `try` block. This avoids moving code around (as a sibling of the caller body or the inlined body), so that should allow more local optimizations after inlining.

### CODE REVIEW

**Comment by tlively** [comment]
My one concern about special casing return calls outside of try blocks is that it might make more work for us in the future if we add more scope-based effects. Unconditionally hoisting the bodies of inlined return calls will always be correct.

**Comment by vouillon** [comment]
> My one concern about special casing return calls outside of try blocks is that it might make more work for us in the future if we add more scope-based effects. Unconditionally hoisting the bodies of inlined return calls will always be correct.

You mean things like dynamic scopes? If you look at my code, you will see that the logic for tracking this is completely separated from the inlining code: it is in a separate class `TryDepthWalker`. So, if another kind of handler is added, it should be fairly easy to find what needs to be changed, and the change should only be a few lines.

File: src/passes/Inlining.cpp [hunk 1]
```diff
@@ -543,7 +564,15 @@ static Expression* doInlining(Module* module,
         builder.makeLocalSet(updater.localMapping[from->getVarIndexBase() + i],
                              LiteralUtils::makeZero(type, *module)));
     }
-    *action.callSite = block;
+    if (call->isReturn /* && !action.insideATry */) {
```
Comment (tlively) [inline src/passes/Inlining.cpp:567]: Why is this commented out? Should this be an assertion instead?

File: src/wasm-traversal.h [hunk 2]
```diff
@@ -537,6 +537,49 @@ struct ExpressionStackWalker : public PostWalker<SubType, VisitorType> {
   }
 };
 
+// Traversal keeping track of try depth
+
+template<typename SubType, typename VisitorType = Visitor<SubType>>
```
Comment (tlively) [inline src/wasm-traversal.h:542]: Can you add a comment saying that if we ever add any other kind of handler scope beyond try-catch, we should handle that here as well?

File: test/lit/passes/inlining_enable-tail-call.wast [hunk 3]
```diff
@@ -884,56 +794,6 @@
  )
 )
 
-(module
- ;; Same as above, but this time the child is not unreachable.
```
Comment (tlively) [inline test/lit/passes/inlining_enable-tail-call.wast:888]: Why remove this test?

**Review by tlively — COMMENTED** [review]
You're right that this should be easy to update in the future, so this change seems good to make. We've lost nearly all our test coverage for the hoisting logic, though. Can you add versions of the tests that use `try_table` to force hoisting?

File: test/lit/passes/inlining_enable-tail-call.wast (hunk 3 above)
Comment (vouillon) [reply test/lit/passes/inlining_enable-tail-call.wast:888]: I think I somehow missed them while rebasing.

**Comment by tlively** [comment]
Looks like this was accidentally closed because the target branch was deleted after it got merged. @vouillon, would you mind opening a copy of this PR against the main branch? Sorry for the extra work.

**Pull request closed without merging**